alerts/cluster-monitoring-operator/PrometheusDuplicateTimestamps.md

```diff
@@ -0,0 +1,83 @@
+
+## Impact
+
+Unwanted samples might be scraped, and desired samples might be dropped.
+Consequently, queries may yield incorrect and unexpected results.
+
+## Diagnosis
+
+1. Determine whether the alert has triggered for the instance of Prometheus used
+   for default cluster monitoring or for the instance that monitors user-defined
+   projects by viewing the alert message's `namespace` label: the namespace for
+   default cluster monitoring is `openshift-monitoring`; the namespace for user
+   workload monitoring is `openshift-user-workload-monitoring`.
+
+2. Review the logs for the affected Prometheus instance:
+
+   ```shell
+   $ NAMESPACE='<value of namespace label from alert>'
+
+   $ oc -n $NAMESPACE logs -l 'app.kubernetes.io/name=prometheus' | \
+   grep 'Error on ingesting samples with different value but same timestamp.*' \
+   | sort | uniq -c | sort -n
+   level=warn ... scrape_pool="the-scrape-pool" target="an-involved-target" \
+   msg="Error on ingesting samples with different value but same timestamp"
+   ```
+   Warning logs similar to the one above should be present.
+
+## Mitigation
+
+If the alert originates from the `openshift-monitoring` namespace, please open a
+support case. If not, proceed as follows:
+
+The alert may fire if a target is duplicated: defined multiple times with
+identical target labels. You can learn if the logs above reveal that the same
+target is defined in different scrape pools.
+
+Also, this may occur if distinct targets across different scrape pools are
+producing the same samples.
+
+In these two cases, the logs above should be sufficient to guide you to the
+right place in the configuration file to find and fix the conflicting targets:
+remove the duplicated targets or ensure that distinct targets are labeled
+uniquely.
+
+If the logs above only revolve around the same scrape pool, this may mean that
+the target exposes duplicated samples. In this scenario, even if they have the
+same value, samples are considered duplicates.
+
+The logs above may not be sufficient anymore to find the specific problematic
+metric within the broken target.
+
+To help identify the concerned metrics, the `debug` log level needs to be
+enabled on the Prometheus instance.
+
+Here is [a guide to change log level] for monitoring components in
+OpenShift.
+
+Then review the new debug logs:
+
+   ```shell
+   $ NAMESPACE='<value of namespace label from alert>'
+
+   $ oc -n $NAMESPACE logs -l 'app=prometheus' | \
+   grep 'Duplicate sample for timestamp.*' | sort | uniq -c | sort -n
+   level=debug ... msg="Duplicate sample for timestamp" \
+   series="a-concerned-series"
+   ```
+
+   The logs should reveal the concerned metrics.
+
+In this case, you will have to fix the metrics exposition in the broken
+target(s) to resolve the problem: make sure they don't expose duplicated
+samples.
+
+After resolving the issue, remember to turn off the `debug` log level.
```

Comment:
I suggest rewriting it so that it is a little easier to read. Here is my suggestion:

Reply:
```

- **The logs reveil one of the following issues:**

   - **The same target is defined in different scrape pools**
   - **Distinct targets across different scrape pools are producing the same samples**

   This happens when the target is duplicated, that is, defined multiple times with identical target labels.

   Proceed with the following steps to fix the issue:

    1. Use the logs to guide you to the right place in the configuration file to find the conflicting targets.
    2. Remove the duplicated targets or ensure that distinct targets are labeled uniquely.

- **The logs only revolve around the same scrape pool:**
   
   This might mean that the target exposes duplicated samples. In this scenario, even if the samples have the same value, they are considered duplicates.

   Proceed with the following steps to fix the issue:

   1. Enable the `debug` log level on the Prometheus instance. See [a guide to change log level] for monitoring components in OpenShift.
   2. Review the new debug logs. The logs should reveal the problematic metrics:
         
      ```shell
      $ NAMESPACE='<value of namespace label from alert>'

      $ oc -n $NAMESPACE logs -l 'app=prometheus' | \
      grep 'Duplicate sample for timestamp.*' | sort | uniq -c | sort -n
      level=debug ... msg="Duplicate sample for timestamp" \
      series="a-concerned-series"
      ```
      In this case, you have to fix the metrics exposition in the broken targets to resolve the problem. Ensure they do not expose duplicated samples.
   3. After resolving the issue, disable the `debug` log level.
```